include/exadg/aero_acoustic/volume_coupling.h

```diff
@@ -0,0 +1,123 @@
+#include <exadg/aero_acoustic/single_field_solvers/fluid.h>
+#include <exadg/aero_acoustic/user_interface/parameters.h>
+
+namespace ExaDG
+{
+namespace AeroAcoustic
+{
+/**
+ * A class that handles the volume coupling between fluid and acoustic.
+ */
+template<int dim, typename Number>
+class VolumeCoupling
+{
+  using VectorType = dealii::LinearAlgebra::distributed::Vector<Number>;
+
+public:
+  void
+  setup(Parameters const &                           parameters_in,
+        std::shared_ptr<SolverAcoustic<dim, Number>> acoustic_solver_in,
+        std::shared_ptr<SolverFluid<dim, Number>>    fluid_solver_in)
+  {
+    parameters      = parameters_in;
+    acoustic_solver = acoustic_solver_in;
+    fluid_solver    = fluid_solver_in;
+
+    acoustic_solver_in->pde_operator->initialize_dof_vector_pressure(source_term_acoustic);
+    fluid_solver_in->pde_operator->initialize_vector_pressure(source_term_fluid);
+
+    // setup the transfer operator
+    if(parameters.fluid_to_acoustic_coupling_strategy ==
+       FluidToAcousticCouplingStrategy::NonNestedMGRestriction)
+    {
+      non_nested_mg.reinit(fluid_solver_in->pde_operator->get_dof_handler_p(),
+                           acoustic_solver_in->pde_operator->get_dof_handler_p(),
+                           *fluid_solver_in->pde_operator->get_mapping(),
+                           *acoustic_solver_in->pde_operator->get_mapping());
+    }
+    else
+    {
+      AssertThrow(false, dealii::ExcMessage("FluidToAcousticCouplingStrategy not implemented."));
+    }
+
+    // setup aeroacoustic source term calculator
+    SourceTermCalculatorData data;
+    data.dof_index_pressure  = fluid_solver_in->pde_operator->get_dof_index_pressure();
+    data.dof_index_velocity  = fluid_solver_in->pde_operator->get_dof_index_velocity();
+    data.quad_index          = fluid_solver_in->pde_operator->get_quad_index_pressure();
+    data.density             = parameters_in.density;
+    data.consider_convection = parameters_in.source_term_with_convection;
+
+    source_term_calculator.setup(fluid_solver_in->pde_operator->get_matrix_free(), data);
+  }
+
+  void
+  fluid_to_acoustic()
+  {
+    if(parameters.fluid_to_acoustic_coupling_strategy ==
+       FluidToAcousticCouplingStrategy::NonNestedMGRestriction)
+    {
+      source_term_calculator.evaluate_integrate(source_term_fluid,
+                                                fluid_solver->time_integrator->get_velocity(),
+                                                fluid_solver->time_integrator->get_pressure(),
+                                                fluid_solver->get_pressure_time_derivative());
+
+      non_nested_mg.restrict_and_add(source_term_acoustic, source_term_fluid);
+    }
+    else
+    {
+      AssertThrow(false, dealii::ExcMessage("FluidToAcousticCouplingStrategy not implemented."));
+    }
+
+    acoustic_solver->pde_operator->set_integrated_rhs(source_term_acoustic);
+  }
```

Comment:
I like that the coupling term calculations happen outside the single field solver implementation (SpatialOperator)!

Reply:
:)